File: tools/tf-types-ts/src/core/bridge-a2a.ts
/**
 * A2A (agent-to-agent) protocol bridge.
 *
 * A2A is the discovery + capability-exchange protocol an AI agent
 * uses when it negotiates with another agent for tool/resource access.
 * Where MCP focuses on tool catalogues, A2A focuses on agent
 * capability advertisements + per-call negotiation.
 *
 * The TrustForge A2A bridge:
 *   - Projects an A2A AgentCard into a TrustForge ActorIdentity.
 *   - Normalises A2A capability names into TrustForge action names
 *     under the `a2a.` prefix (mirror of the MCP normalisation).
 *   - Surfaces every advertised capability as a Capability the
 *     AgentGuard can authorize per call.
 *
 * The Rust mirror lives at `crates/tf-types/src/bridge_a2a.rs`.
 */

import type { ActionName, ActorId, Capability } from "../generated/_common.js";
import type { ActorIdentity } from "../generated/actor-identity.js";
import { type Bridge, type BridgeKind, BridgeFailure } from "./bridges.js";

export interface A2AAgentCard {
  /** Canonical A2A agent identifier (DNS-style: `agent.example.com`). */
  agent_id: string;
  /** Human-readable display name. */
  display_name?: string;
  /** Public key the remote agent uses to sign A2A messages. */
  public_key_b64?: string;
  /** Algorithm the public key uses. Default ed25519. */
  public_key_algorithm?: string;
  /** Capability advertisements. Each entry's `name` is the A2A
   *  capability id; the bridge normalises it into a TrustForge
   *  action name. */
  capabilities: Array<{
    name: string;
    description?: string;
    risk?: Capability["risk"];
  }>;
  /** Trust domain the agent runs under. */
  trust_domain: string;
}

export interface A2ABridgeConfig {
  bridgeId: string;
  trustDomain: string;
  /** Default risk class for capabilities the agent didn't specify one for. */
  defaultRisk?: Capability["risk"];
}

export interface A2ABridgeAcceptResult {
  identity: ActorIdentity;
  capabilities: Capability[];
  source: "a2a-agent-card";
}

const ACTION_NAME_RE = /^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]*)+$/;

/**
 * Normalise an A2A capability name into a TrustForge action name. The
 * rules mirror MCP's normaliseToolName but always prefix `a2a.`:
 *   - lowercase
 *   - non-alphanumeric runs collapse to a single `_`
 *   - leading / trailing `_` stripped
 *   - if the result has no `.`, prepend `a2a.`
 */
export function a2aNormaliseCapability(name: string, prefix?: string): string {
  const scrubbed = name
    .replace(/[^A-Za-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "")
    .toLowerCase();
  const withPrefix = prefix ? `${prefix}.${scrubbed}` : scrubbed;
  return withPrefix.includes(".") ? withPrefix : `a2a.${withPrefix}`;
}

export class A2ABridge implements Bridge {
  readonly kind: BridgeKind = "a2a";
  readonly bridgeId: string;
  readonly trustDomain: string;
  private readonly cfg: A2ABridgeConfig;

  constructor(cfg: A2ABridgeConfig) {
    this.bridgeId = cfg.bridgeId;
    this.trustDomain = cfg.trustDomain;
    this.cfg = cfg;
  }

  /** Project an A2A AgentCard into a TrustForge ActorIdentity +
   *  capability list. Throws BridgeFailure on malformed input. */
  acceptAgentCard(card: A2AAgentCard): A2ABridgeAcceptResult {
    if (!card.agent_id || !card.trust_domain || !Array.isArray(card.capabilities)) {
      throw new BridgeFailure({ code: "invalid-input", message: "A2A AgentCard missing required fields" });
    }
    const actor: ActorId = `tf:actor:agent:${card.trust_domain}/${card.agent_id}`;
    const algo = card.public_key_algorithm ?? "ed25519";
    const identity: ActorIdentity = {
      identity_version: "1",
      actor_id: actor,
      actor_type: "agent",
      public_keys: card.public_key_b64
        ? [
            {
              key_id: "a2a-agent-card",
              algorithm: algo,
              public_key: card.public_key_b64,
              purpose: "signing",
            },
          ]
        : [
            {
              key_id: "a2a-agent-card",
              algorithm: "external-attestation",
              // No key shipped on this AgentCard — mark explicitly so
              // callers don't treat the entry as a verified ed25519
              // key.
              public_key: `agent-card:${card.agent_id}`,
              purpose: "attestation",
            },
          ],
      trust_levels: ["T2"],
      authority_roots: [{ kind: "federation", id: card.trust_domain }],
      valid_from: new Date().toISOString(),
    };
    const capabilities: Capability[] = card.capabilities.map((c) => {
      const action = a2aNormaliseCapability(c.name) as ActionName;
      // Sanity-check the normalised name matches the schema regex; an
      // unrepresentable A2A name surfaces loudly rather than silently
      // breaking AgentGuard.
      if (!ACTION_NAME_RE.test(action)) {
        throw new BridgeFailure({
          code: "rejected",
          message: `A2A capability ${c.name} does not normalise to a valid action name (got ${action})`,
        });
      }
      return { name: action, risk: c.risk ?? this.cfg.defaultRisk ?? "R2" };
    });
    return { identity, capabilities, source: "a2a-agent-card" };
  }
}

File: crates/tf-types/src/bridge_a2a.rs
//! A2A (agent-to-agent) protocol bridge — mirror of TS bridge-a2a.ts.

use crate::bridges::{Bridge, BridgeError, BridgeKind};

#[derive(Clone, Debug)]
pub struct A2AAgentCard {
    pub agent_id: String,
    pub display_name: Option<String>,
    pub public_key_b64: Option<String>,
    pub public_key_algorithm: Option<String>,
    pub capabilities: Vec<A2ACapability>,
    pub trust_domain: String,
}

#[derive(Clone, Debug)]
pub struct A2ACapability {
    pub name: String,
    pub description: Option<String>,
    pub risk: Option<String>,
}

#[derive(Clone, Debug)]
pub struct A2ABridgeConfig {
    pub bridge_id: String,
    pub trust_domain: String,
    pub default_risk: Option<String>,
}

pub struct A2ABridge {
    cfg: A2ABridgeConfig,
}

impl Bridge for A2ABridge {
    fn bridge_id(&self) -> &str {
        &self.cfg.bridge_id
    }
    fn kind(&self) -> BridgeKind {
        BridgeKind::A2a
    }
    fn trust_domain(&self) -> &str {
        &self.cfg.trust_domain
    }
}

#[derive(Clone, Debug)]
pub struct ProjectedActor {
    pub actor_id: String,
    pub algorithm: String,
    pub public_key: String,
    pub capabilities: Vec<(String, String)>, // (action_name, risk)
}

impl A2ABridge {
    pub fn new(cfg: A2ABridgeConfig) -> Self {
        Self { cfg }
    }

    pub fn accept_agent_card(&self, card: &A2AAgentCard) -> Result<ProjectedActor, BridgeError> {
        if card.agent_id.is_empty() || card.trust_domain.is_empty() {
            return Err(BridgeError::InvalidInput(
                "AgentCard missing agent_id or trust_domain".into(),
            ));
        }
        let actor_id = format!("tf:actor:agent:{}/{}", card.trust_domain, card.agent_id);
        let (algorithm, public_key) = match (&card.public_key_b64, &card.public_key_algorithm) {
            (Some(pk), Some(alg)) => (alg.clone(), pk.clone()),
            (Some(pk), None) => ("ed25519".to_string(), pk.clone()),
            _ => (
                "external-attestation".to_string(),
                format!("agent-card:{}", card.agent_id),
            ),
        };
        let mut caps = Vec::with_capacity(card.capabilities.len());
        for c in &card.capabilities {
            let action = a2a_normalise_capability(&c.name, None);
            if !is_valid_action_name(&action) {
                return Err(BridgeError::Rejected(format!(
                    "A2A capability {} does not normalise to a valid action name (got {})",
                    c.name, action
                )));
            }
            let risk = c
                .risk
                .clone()
                .or_else(|| self.cfg.default_risk.clone())
                .unwrap_or_else(|| "R2".to_string());
            caps.push((action, risk));
        }
        Ok(ProjectedActor {
            actor_id,
            algorithm,
            public_key,
            capabilities: caps,
        })
    }
}

/// Mirror of TS `a2aNormaliseCapability`: lowercase, non-alphanumeric
/// runs collapse to `_`, leading/trailing `_` stripped, prepend `a2a.`
/// if no dot is present.
pub fn a2a_normalise_capability(name: &str, prefix: Option<&str>) -> String {
    let mut buf = String::with_capacity(name.len());
    let mut last_underscore = false;
    for c in name.chars() {
        if c.is_ascii_alphanumeric() {
            buf.push(c.to_ascii_lowercase());
            last_underscore = false;
        } else if !last_underscore {
            buf.push('_');
            last_underscore = true;
        }
    }
    let scrubbed = buf.trim_matches('_').to_string();
    let with_prefix = match prefix {
        Some(p) => format!("{}.{}", p, scrubbed),
        None => scrubbed,
    };
    if with_prefix.contains('.') {
        with_prefix
    } else {
        format!("a2a.{}", with_prefix)
    }
}

fn is_valid_action_name(s: &str) -> bool {
    let mut segs = s.split('.');
    let first = match segs.next() {
        Some(x) => x,
        None => return false,
    };
    if !is_valid_action_segment(first) {
        return false;
    }
    let mut count = 1;
    for seg in segs {
        if !is_valid_action_segment(seg) {
            return false;
        }
        count += 1;
    }
    count >= 2
}

fn is_valid_action_segment(s: &str) -> bool {
    let mut chars = s.chars();
    let first = match chars.next() {
        Some(c) => c,
        None => return false,
    };
    if !first.is_ascii_lowercase() {
        return false;
    }
    chars.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalise_basic() {
        // dot is non-alphanumeric → collapsed to `_`, then `a2a.` prefix added.
        assert_eq!(a2a_normalise_capability("filesystem.read", None), "a2a.filesystem_read");
        assert_eq!(a2a_normalise_capability("ping", None), "a2a.ping");
        assert_eq!(a2a_normalise_capability("Read File!", None), "a2a.read_file");
        assert_eq!(a2a_normalise_capability("system-info", Some("tools")), "tools.system_info");
    }

    #[test]
    fn accept_agent_card_round_trip() {
        let bridge = A2ABridge::new(A2ABridgeConfig {
            bridge_id: "tf-a2a".into(),
            trust_domain: "example.com".into(),
            default_risk: Some("R2".into()),
        });
        let card = A2AAgentCard {
            agent_id: "code-helper".into(),
            display_name: None,
            public_key_b64: Some("AAAA".into()),
            public_key_algorithm: Some("ed25519".into()),
            capabilities: vec![A2ACapability {
                name: "fs.read".into(),
                description: None,
                risk: None,
            }],
            trust_domain: "example.com".into(),
        };
        let p = bridge.accept_agent_card(&card).expect("project");
        assert_eq!(p.actor_id, "tf:actor:agent:example.com/code-helper");
        assert_eq!(p.capabilities[0].0, "a2a.fs_read");
        assert_eq!(p.capabilities[0].1, "R2");
    }

    #[test]
    fn missing_agent_id_rejected() {
        let bridge = A2ABridge::new(A2ABridgeConfig {
            bridge_id: "tf-a2a".into(),
            trust_domain: "example.com".into(),
            default_risk: None,
        });
        let card = A2AAgentCard {
            agent_id: "".into(),
            display_name: None,
            public_key_b64: None,
            public_key_algorithm: None,
            capabilities: vec![],
            trust_domain: "example.com".into(),
        };
        assert!(bridge.accept_agent_card(&card).is_err());
    }
}
